tools/dataImporter.php

```diff
@@ -0,0 +1,255 @@
+
+/*
+ * Expected indices of parameters passed to this script via $argv
+ *
+ * @var int
+ */
+const MAPPING_ARG_INDEX = 1;
+const DATA_ARG_INDEX = 2;
+
+/**
+ * These constants represent the indices of columns within the mapping file.
+ * The file passed must have the headers organized in this order. 
+ * The names of the columns in the csv file don't matter as long as they are formatted
+ * like this.
+ *
+ *
+ * Format:
+ *      OLD_PSCID,NEW_PSCID,NEW_CANDID
+ *
+ * @var int
+ */
+const OLD_PSCID = 0;
+const NEW_PSCID = 1;
+const NEW_CANDID = 2;
+
+/**
+ * The DATA file passed must have the headers organized in this order. This
+ * file is generated by running the dataImporter tool. 
+ * The CSV header COLUMN_NAME MUST match the source and destination column 
+ * name in the MySQL database as this is how the script determines where the 
+ * data should be updated.
+ * COLUMN_NAME contains the data from the source database.
+ *
+ * Format:
+ *      OLD_PSCID,COLUMN_NAME
+ *
+ * @var int
+ */
+const COLUMN_NAME = 1;
+
+/**
+ * These constants are used to make SQL command output more clear.
+ *
+ * @var int
+ */
+const SET_COLUMN = 0;
+const SET_NEWVALUE = 1;
+const WHERE_COLUMN = 0;
+const WHERE_OLDVALUE = 1;
+
+require_once "generic_includes.php";
+
+$usage = <<<USAGE
+Usage: php {$argv[0]} <mapping.csv> <data.csv>
+<mapping.csv>   A CSV file containg columns with NEWPSCID, 
+                OLDPSCID, and NEWCANDID
+<data.csv>      A CSV file containing columns with OLDPSCID,DATA
+USAGE;
+
+
+/* BEGIN SCRIPT */
+
+// Die if not enough arguments
+if (count($argv) < NUM_ARGS_REQUIRED) {
+    die ($usage);
+}
+
+// Get data from CSV files.
+$mappingRows = populateArrayFromCSV($argv[MAPPING_ARG_INDEX]);
+$dataRows = populateArrayFromCSV($argv[DATA_ARG_INDEX]);
+
+// Creating this array allows referencing the columns by index. This allows the
+// use of name constants to access the data from within the CSV files instead
+// of hard-coding column names into this script.
+$mappingHeaders = array_keys($mappingRows[0]);
+$dataHeaders = array_keys($dataRows[0]);
+
+// Find all the candidates present in both the data file and the mapping file by
+// doing a set intersection on the list of old PSCIDs. 
+// It's more efficient to do this once before iterating over every data row 
+// rather than checking if a CandID exists in both arrays on each iteration of 
+// the data.csv file.
+$oldPSCIDsInMappingFile = array();
+$oldPSCIDsInDataFile = array();
+foreach($mappingRows as $row) {
+    $oldPSCIDsInMappingFile[] = $row[$mappingHeaders[OLD_PSCID]];
+}
+foreach($dataRows as $row) {
+    $oldPSCIDsInDataFile[] = $row[$dataHeaders[OLD_PSCID]];
+}
+// array_intersect preserves keys but we don't want them.
+$sharedCandidates = array_values(
+    array_intersect($oldPSCIDsInMappingFile, $oldPSCIDsInDataFile)
+);
+
+// Build a one-to-one mapping of old to new PSCIDs with the old IDs as the key.
+$PSCIDMapping = array();
+foreach($mappingRows as $row) {
```

Comment:
How about iterating over `$sharedCandidates` instead? (haven't thought about whether this approach would be more extensible or not though)

Reply:
Yeah that probably makes more sense

Reply:
You could probably iterate over `$sharedCandidates` on line 147(https://github.com/aces/Loris/pull/4417/files#diff-4b03c0429757434e7d7c0d9634690cceR147) too and have less checks and directly access the data array with the help of `$PSCIDMapping`. 
But yeah both of these are just bonus enhancements. This is a nice prototype! Good job.

Reply:
Actually I think this way works more sense just given the way the data is organized. The function I'm using to put the CSV into an array creates an array like this:
```
[0] => Array (
     [PSCID] => MON123,
     [COLUMN] => some_field
)
...
```
So there's no way to directly look up on PSCID (using `$sharedCandidates`).

The issue comes with line 152 -- you can't directly grab a field like that given the array structure. You'd have to search through `$dataRows` every time which makes the loop O(n^2) whereas the preprocessing above line 147 makes the foreach loop O(n)... I think

Not sure if that makes sense, I am a little braindead right now